File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/network/OkHttpNetworkConnection.java
package com.snowplowanalytics.snowplow.network;

import android.content.Context;
import android.net.TrafficStats;
import android.net.Uri;
import android.os.Build;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.snowplowanalytics.snowplow.internal.emitter.Executor;
import com.snowplowanalytics.snowplow.internal.emitter.TLSArguments;
import com.snowplowanalytics.snowplow.internal.emitter.TLSVersion;
import com.snowplowanalytics.snowplow.tracker.BuildConfig;
import com.snowplowanalytics.snowplow.internal.constants.TrackerConstants;
import com.snowplowanalytics.snowplow.internal.tracker.Logger;

import java.io.IOException;
import java.util.ArrayList;
import java.util.EnumSet;
import java.util.HashMap;
import java.util.List;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

import okhttp3.Cookie;
import okhttp3.CookieJar;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.RequestBody;
import okhttp3.Response;

import static com.snowplowanalytics.snowplow.network.HttpMethod.GET;
import static com.snowplowanalytics.snowplow.network.HttpMethod.POST;

/**
 * Components in charge to send events to the collector.
 * It uses OkHttp as Http client.
 */
public class OkHttpNetworkConnection implements NetworkConnection {
    private final String TAG = OkHttpNetworkConnection.class.getSimpleName();

    private static final int TRAFFIC_STATS_TAG = 1;
    private static final String DEFAULT_USER_AGENT = String.format("snowplow/%s android/%s", BuildConfig.TRACKER_LABEL, Build.VERSION.RELEASE);
    private final MediaType JSON = MediaType.parse(TrackerConstants.POST_CONTENT_TYPE);

    private final String uri;
    private final Protocol protocol;
    private final HttpMethod httpMethod;
    private final int emitTimeout;
    private final String customPostPath;
    private final boolean serverAnonymisation;

    private OkHttpClient client;
    private Uri.Builder uriBuilder;

    /**
     * Builder for the OkHttpNetworkConnection.
     */
    public static class OkHttpNetworkConnectionBuilder {
        final String uri; // Required
        Context context; // Required
        HttpMethod httpMethod = POST; // Optional
        EnumSet<TLSVersion> tlsVersions = EnumSet.of(TLSVersion.TLSv1_2); // Optional
        private int emitTimeout = 5; // Optional
        OkHttpClient client = null; //Optional
        CookieJar cookieJar = null; // Optional
        String customPostPath = null; //Optional
        boolean serverAnonymisation = false; // Optional

        /**
         * @param uri The uri of the collector
         */
        public OkHttpNetworkConnectionBuilder(@NonNull String uri, @NonNull Context context) {
            this.uri = uri;
            this.context = context;
        }

        /**
         * @param httpMethod The method by which requests are emitted
         * @return itself
         */
        @NonNull
        public OkHttpNetworkConnectionBuilder method(@NonNull HttpMethod httpMethod) {
            this.httpMethod = httpMethod;
            return this;
        }

        /**
         * @param version the TLS version allowed for requests
         * @return itself
         */
        @NonNull
        public OkHttpNetworkConnectionBuilder tls(@NonNull TLSVersion version) {
            this.tlsVersions = EnumSet.of(version);
            return this;
        }

        /**
         * @param versions the TLS versions allowed for requests
         * @return itself
         */
        @NonNull
        public OkHttpNetworkConnectionBuilder tls(@NonNull EnumSet<TLSVersion> versions) {
            this.tlsVersions = versions;
            return this;
        }

        /**
         * @param emitTimeout The maximum timeout for emitting events. If emit time exceeds this value
         *                    TimeOutException will be thrown
         * @return itself
         */
        @NonNull
        public OkHttpNetworkConnectionBuilder emitTimeout(int emitTimeout){
            this.emitTimeout = emitTimeout;
            return this;
        }

        /**
         * @param client An OkHttp client that will be used in the emitter, you can provide your
         *               own if you want to share your Singleton client's interceptors, connection pool etc..
         *               ,otherwise a new one is created.
         * @return itself
         */
        @NonNull
        public OkHttpNetworkConnectionBuilder client(@Nullable OkHttpClient client) {
            this.client = client;
            return this;
        }

        /**
         * @param cookieJar An OkHttp cookie jar to override the default cookie jar that stores
         *                  cookies in SharedPreferences. The cookie jar will be ignored in case
         *                  custom `client` is configured.
         * @return itself
         */
        @NonNull
        public OkHttpNetworkConnectionBuilder cookieJar(@Nullable CookieJar cookieJar) {
            this.cookieJar = cookieJar;
            return this;
        }

        /**
         * @param customPostPath A custom path that is used on the endpoint to send requests.
         * @return itself
         */
        @NonNull
        public OkHttpNetworkConnectionBuilder customPostPath(@Nullable String customPostPath) {
            this.customPostPath = customPostPath;
            return this;
        }

        /**
         * @param serverAnonymisation whether to anonymise server-side user identifiers including the `network_userid` and `user_ipaddress`
         * @return itself
         */
        @NonNull
        public OkHttpNetworkConnectionBuilder serverAnonymisation(@Nullable boolean serverAnonymisation) {
            this.serverAnonymisation = serverAnonymisation;
            return this;
        }

        /**
         * Creates a new OkHttpNetworkConnection
         *
         * @return a new OkHttpNetworkConnection object
         */
        @NonNull
        public OkHttpNetworkConnection build() {
            return new OkHttpNetworkConnection(this);
        }
    }

    private OkHttpNetworkConnection(OkHttpNetworkConnectionBuilder builder) {
        // Decode uri to extract protocol
        String tempUri = builder.uri;
        Uri url = Uri.parse(builder.uri);
        Protocol tempProtocol = Protocol.HTTPS;
        if (url.getScheme() == null) {
            tempUri = "https://" + builder.uri;
        } else {
            switch (url.getScheme()) {
                case "https":
                    break;
                case "http":
                    tempProtocol = Protocol.HTTP;
                    break;
                default:
                    tempUri = "https://" + builder.uri;
            }
        }

        // Configure
        uri = tempUri;
        protocol = tempProtocol;
        httpMethod = builder.httpMethod;
        emitTimeout = builder.emitTimeout;
        customPostPath = builder.customPostPath;
        serverAnonymisation = builder.serverAnonymisation;

        TLSArguments tlsArguments = new TLSArguments(builder.tlsVersions);
        String protocolString = protocol == Protocol.HTTP ? "http://" : "https://";
        uriBuilder = Uri.parse(uri).buildUpon();

        if (httpMethod == GET) {
            uriBuilder.appendPath("i");
        } else if (this.customPostPath == null) {
            uriBuilder.appendEncodedPath(TrackerConstants.PROTOCOL_VENDOR + "/" +
                    TrackerConstants.PROTOCOL_VERSION);
        } else {
            uriBuilder.appendEncodedPath(this.customPostPath);
        }

        // Configure with external OkHttpClient
        if (builder.client == null) {
            client = new OkHttpClient.Builder()
                    .sslSocketFactory(tlsArguments.getSslSocketFactory(), tlsArguments.getTrustManager())
                    .connectTimeout(15, TimeUnit.SECONDS)
                    .readTimeout(15, TimeUnit.SECONDS)
                    .cookieJar(builder.cookieJar == null ? new CollectorCookieJar(builder.context) : builder.cookieJar)
                    .build();
        } else {
            client = builder.client;
        }
    }

    @Override
    @NonNull
    public HttpMethod getHttpMethod() {
        return httpMethod;
    }

    @NonNull
    @Override
    public Uri getUri() {
        return uriBuilder.clearQuery().build();
    }

    @NonNull
    @Override
    public List<RequestResult> sendRequests(@NonNull List<Request> requests) {
        List<Future> futures = new ArrayList<>();
        List<RequestResult> results = new ArrayList<>();

        // Start all requests in the ThreadPool
        for (Request request : requests) {
            String userAgent = request.customUserAgent != null ? request.customUserAgent : DEFAULT_USER_AGENT;

            okhttp3.Request okHttpRequest = httpMethod == HttpMethod.GET
                    ? buildGetRequest(request, userAgent)
                    : buildPostRequest(request, userAgent);

            futures.add(Executor.futureCallable(getRequestCallable(okHttpRequest)));
        }

        Logger.d(TAG, "Request Futures: %s", futures.size());

        // Get results of futures
        // - Wait up to emitTimeout seconds for the request
        for (int i = 0; i < futures.size(); i++) {
            int code = -1;

            try {
                code = (int) futures.get(i).get(emitTimeout, TimeUnit.SECONDS);
            } catch (InterruptedException ie) {
                Logger.e(TAG, "Request Future was interrupted: %s", ie.getMessage());
            } catch (ExecutionException ee) {
                Logger.e(TAG, "Request Future failed: %s", ee.getMessage());
            } catch (TimeoutException te) {
                Logger.e(TAG, "Request Future had a timeout: %s", te.getMessage());
            }

            Request request = requests.get(i);
            List<Long> eventIds = request.emitterEventIds;
            results.add(new RequestResult(code, request.oversize, eventIds));
            if (request.oversize) {
                Logger.track(TAG, "Request is oversized for emitter event IDs: %s", eventIds.toString());
            }
        }
        return results;
    }

    /**
     * Builds an OkHttp GET request which is ready
     * to be executed.
     * @param request The request where to get the payload to be sent.
     * @param userAgent The user-agent used during the transmission to the collector.
     * @return An OkHttp request object.
     */
    private okhttp3.Request buildGetRequest(Request request, String userAgent) {
        // Clear the previous query
        uriBuilder.clearQuery();

        // Build the request query
        HashMap hashMap = (HashMap) request.payload.getMap();

        for (String key : (Iterable<String>) hashMap.keySet()) {
            String value = (String) hashMap.get(key);
            uriBuilder.appendQueryParameter(key, value);
        }

        // Build the request
        String reqUrl = uriBuilder.build().toString();
        okhttp3.Request.Builder builder = new okhttp3.Request.Builder()
                .url(reqUrl)
                .header("User-Agent", userAgent)
                .get();
        if (serverAnonymisation) {
            builder.header("SP-Anonymous", "*");
        }
        return builder.build();
    }

    /**
     * Builds an OkHttp POST request which is ready
     * to be executed.
     * @param request The request where to get the payload to be sent.
     * @param userAgent The user-agent used during the transmission to the collector.
     * @return An OkHttp request object.
     */
    private okhttp3.Request buildPostRequest(Request request, String userAgent) {
        String reqUrl = uriBuilder.build().toString();
        RequestBody reqBody = RequestBody.create(JSON, request.payload.toString());
        okhttp3.Request.Builder builder = new okhttp3.Request.Builder()
                .url(reqUrl)
                .header("User-Agent", userAgent)
                .post(reqBody);
        if (serverAnonymisation) {
            builder.header("SP-Anonymous", "*");
        }
        return builder.build();
    }

    /**
     * Returns a Callable Request Send
     *
     * @param request the request to be
     *                sent
     * @return the new Callable object
     */
    private Callable<Integer> getRequestCallable(final okhttp3.Request request) {
        return () -> requestSender(request);
    }

    /**
     * The function responsible for actually sending
     * the request to the collector.
     *
     * @param request The request to be sent
     * @return a RequestResult
     */
    private int requestSender(okhttp3.Request request) {
        try {
            Logger.v(TAG, "Sending request: %s", request);
            TrafficStats.setThreadStatsTag(TRAFFIC_STATS_TAG);
            Response resp = client.newCall(request).execute();
            int code = resp.code();
            resp.body().close();

            return code;
        } catch (IOException e) {
            Logger.e(TAG, "Request sending failed: %s", e.toString());
            return -1;
        }
    }

    /**
     * Returns truth on if the request
     * was sent successfully.
     *
     * @param code the response code
     * @return the truth as to the success
     */
    private boolean isSuccessfulSend(int code) {
        return code >= 200 && code < 300;
    }
}

File: snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/network/OkHttpNetworkConnection.kt
/*
 * Copyright (c) 2015-2023 Snowplow Analytics Ltd. All rights reserved.
 *
 * This program is licensed to you under the Apache License Version 2.0,
 * and you may not use this file except in compliance with the Apache License Version 2.0.
 * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the Apache License Version 2.0 is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
 */
package com.snowplowanalytics.snowplow.network

import android.content.Context
import android.net.TrafficStats
import android.net.Uri
import android.os.Build

import com.snowplowanalytics.core.constants.TrackerConstants
import com.snowplowanalytics.core.emitter.EmitterDefaults
import com.snowplowanalytics.core.emitter.Executor
import com.snowplowanalytics.core.emitter.TLSArguments
import com.snowplowanalytics.core.emitter.TLSVersion
import com.snowplowanalytics.core.tracker.Logger
import com.snowplowanalytics.snowplow.tracker.BuildConfig

import okhttp3.CookieJar
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.OkHttpClient
import okhttp3.RequestBody.Companion.toRequestBody

import java.io.IOException
import java.util.*
import java.util.concurrent.*

/**
 * Components in charge to send events to the collector.
 * It uses OkHttp as Http client.
 */
class OkHttpNetworkConnection private constructor(builder: OkHttpNetworkConnectionBuilder) :
    NetworkConnection {
    private val TAG = OkHttpNetworkConnection::class.java.simpleName
    private val JSON = TrackerConstants.POST_CONTENT_TYPE.toMediaTypeOrNull()
    
    private val networkUri: String
    private val protocol: Protocol
    override val httpMethod: HttpMethod
    private val emitTimeout: Int
    private val customPostPath: String?
    private val serverAnonymisation: Boolean
    private val requestHeaders: Map<String, String>?
    private var client: OkHttpClient? = null
    private val uriBuilder: Uri.Builder
    override val uri: Uri
        get() = uriBuilder.clearQuery().build()
    
    /**
     * Builder for the OkHttpNetworkConnection.
     * @param uri The uri of the collector
     */
    class OkHttpNetworkConnectionBuilder(
        val uri: String,
        val context: Context
    ) {
        var httpMethod = EmitterDefaults.httpMethod // Optional
        var tlsVersions: EnumSet<TLSVersion> = EmitterDefaults.tlsVersions // Optional
        var emitTimeout = EmitterDefaults.emitTimeout // Optional
        var client: OkHttpClient? = null // Optional
        var cookieJar: CookieJar? = null // Optional
        var customPostPath: String? = null //Optional
        var serverAnonymisation = EmitterDefaults.serverAnonymisation // Optional
        var requestHeaders: Map<String, String>? = null // Optional

        /**
         * GET or POST.
         * @param httpMethod The method by which requests are emitted
         * @return itself
         */
        fun method(httpMethod: HttpMethod): OkHttpNetworkConnectionBuilder {
            this.httpMethod = httpMethod
            return this
        }

        /**
         * This configuration option is not published in the 
         * [NetworkConfiguration](com.snowplowanalytics.snowplow.configuration.NetworkConfiguration) class.
         * Create an Emitter and Tracker directly, not via the 
         * [Snowplow](com.snowplowanalytics.snowplow.Snowplow) interface, to configure tlsVersions.
         * @param version the TLS version allowed for requests
         * @return itself
         */
        fun tls(version: TLSVersion): OkHttpNetworkConnectionBuilder {
            tlsVersions = EnumSet.of(version)
            return this
        }

        /**
         * This configuration option is not published in the
         * [NetworkConfiguration](com.snowplowanalytics.snowplow.configuration.NetworkConfiguration) class.
         * Create an Emitter and Tracker directly, not via the
         * [Snowplow](com.snowplowanalytics.snowplow.Snowplow) interface, to configure tlsVersions.
         * @param versions the TLS versions allowed for requests
         * @return itself
         */
        fun tls(versions: EnumSet<TLSVersion>): OkHttpNetworkConnectionBuilder {
            tlsVersions = versions
            return this
        }

        /**
         * Maximum time allowed for a request. If emit time exceeds this value a
         * TimeOutException will be thrown, which is caught and logged.
         * @param emitTimeout The maximum timeout for emitting events. 
         * @return itself
         */
        fun emitTimeout(emitTimeout: Int): OkHttpNetworkConnectionBuilder {
            this.emitTimeout = emitTimeout
            return this
        }

        /**
         * A custom OkHttp client that will be used in the emitter. Provide your
         * own if you want to share your Singleton client's interceptors, connection pool etc.,
         * otherwise a new one is created. as part of the OkHttpNetworkConnection instantiation.
         * @param client An OkHttp client
         * @return itself
         */
        fun client(client: OkHttpClient?): OkHttpNetworkConnectionBuilder {
            this.client = client
            return this
        }

        /**
         * An OkHttp cookie jar to override the default [CollectorCookieJar](com.snowplowanalytics.snowplow.network.CollectorCookieJar) that stores
         * cookies in SharedPreferences. The cookie jar will be ignored when custom [client] is configured.
         * @param cookieJar OkHttp CookieJar
         * @return itself
         */
        fun cookieJar(cookieJar: CookieJar?): OkHttpNetworkConnectionBuilder {
            this.cookieJar = cookieJar
            return this
        }

        /**
         * A custom path which will be added to the endpoint URL to specify the
         * complete URL of the event collector when paired with the POST method. The default path is
         * "com.snowplowanalytics/snowplow/tp2".
         * The collector must be configured to accept the custom path.
         * @param customPostPath A custom path that is used on the endpoint to send requests.
         * @return itself
         */
        fun customPostPath(customPostPath: String?): OkHttpNetworkConnectionBuilder {
            this.customPostPath = customPostPath
            return this
        }

        /**
         * Whether to anonymise server-side user identifiers including the `network_userid` and `user_ipaddress`.
         * This configuration option is not published in the
         * [NetworkConfiguration](com.snowplowanalytics.snowplow.configuration.NetworkConfiguration) class.
         * Instead, configure it via [EmitterConfiguration](com.snowplowanalytics.snowplow.configuration.EmitterConfiguration).
         * @param serverAnonymisation whether to anonymise
         * @return itself
         */
        fun serverAnonymisation(serverAnonymisation: Boolean): OkHttpNetworkConnectionBuilder {
            this.serverAnonymisation = serverAnonymisation
            return this
        }

        /**
         * A map of custom HTTP headers to add to the request.
         */
        fun requestHeaders(requestHeaders: Map<String, String>?): OkHttpNetworkConnectionBuilder {
            this.requestHeaders = requestHeaders
            return this
        }

        /**
         * Creates a new OkHttpNetworkConnection
         *
         * @return a new OkHttpNetworkConnection object
         */
        fun build(): OkHttpNetworkConnection {
            return OkHttpNetworkConnection(this)
        }
    }

    init {
        // Decode uri to extract protocol
        var tempUri = builder.uri
        val url = Uri.parse(builder.uri)
        var tempProtocol = Protocol.HTTPS
        if (url.scheme == null) {
            tempUri = "https://" + builder.uri
        } else {
            when (url.scheme) {
                "https" -> {}
                "http" -> tempProtocol = Protocol.HTTP
                else -> tempUri = "https://" + builder.uri
            }
        }

        // Configure
        networkUri = tempUri
        protocol = tempProtocol
        httpMethod = builder.httpMethod
        emitTimeout = builder.emitTimeout
        customPostPath = builder.customPostPath
        serverAnonymisation = builder.serverAnonymisation
        requestHeaders = builder.requestHeaders
        
        val tlsArguments = TLSArguments(builder.tlsVersions)
        uriBuilder = Uri.parse(networkUri).buildUpon()
        
        if (httpMethod == HttpMethod.GET) {
            uriBuilder.appendPath("i")
        } else if (customPostPath == null) {
            uriBuilder.appendEncodedPath(
                TrackerConstants.PROTOCOL_VENDOR + "/" +
                        TrackerConstants.PROTOCOL_VERSION
            )
        } else {
            uriBuilder.appendEncodedPath(customPostPath)
        }

        // Configure with external OkHttpClient
        client = if (builder.client == null) {
            OkHttpClient.Builder()
                .sslSocketFactory(tlsArguments.sslSocketFactory, tlsArguments.trustManager)
                .connectTimeout(15, TimeUnit.SECONDS)
                .readTimeout(15, TimeUnit.SECONDS)
                .cookieJar(builder.cookieJar ?: CollectorCookieJar(builder.context))
                .build()
        } else {
            builder.client
        }
    }

    override fun sendRequests(requests: List<Request>): List<RequestResult> {
        val futures: MutableList<Future<*>> = ArrayList()
        val results: MutableList<RequestResult> = ArrayList()

        // Start all requests in the ThreadPool
        for (request in requests) {
            val userAgent = request.customUserAgent ?: DEFAULT_USER_AGENT
            val okHttpRequest = if (httpMethod == HttpMethod.GET) buildGetRequest(
                request,
                userAgent
            ) else buildPostRequest(request, userAgent)
            futures.add(Executor.futureCallable(getRequestCallable(okHttpRequest)))
        }
        Logger.d(TAG, "Request Futures: %s", futures.size)

        // Get results of futures
        // - Wait up to emitTimeout seconds for the request
        for (i in futures.indices) {
            var code: Int = -1
            
            try {
                val tempCode = futures[i][emitTimeout.toLong(), TimeUnit.SECONDS] as? Int
                tempCode?.let { code = it }
            } catch (ie: InterruptedException) {
                Logger.e(TAG, "Request Future was interrupted: %s", ie.message)
            } catch (ee: ExecutionException) {
                Logger.e(TAG, "Request Future failed: %s", ee.message)
            } catch (te: TimeoutException) {
                Logger.e(TAG, "Request Future had a timeout: %s", te.message)
            }
            
            val request = requests[i]
            val eventIds = request.emitterEventIds
            results.add(RequestResult(code, request.oversize, eventIds))
            if (request.oversize) {
                Logger.track(
                    TAG,
                    "Request is oversized for emitter event IDs: %s",
                    eventIds.toString()
                )
            }
        }
        return results
    }

    /**
     * Builds an OkHttp GET request which is ready
     * to be executed.
     * @param request The request where to get the payload to be sent.
     * @param userAgent The user-agent used during the transmission to the collector.
     * @return An OkHttp request object.
     */
    private fun buildGetRequest(request: Request, userAgent: String): okhttp3.Request {
        // Clear the previous query
        uriBuilder.clearQuery()

        // Build the request query
        val hashMap = request.payload.map
        for (key in hashMap.keys) {
            val value = hashMap[key] as? String?
            uriBuilder.appendQueryParameter(key, value)
        }

        // Build the request
        val reqUrl = uriBuilder.build().toString()
        val builder = okhttp3.Request.Builder()
            .url(reqUrl)
            .header("User-Agent", userAgent)
            .get()
        if (serverAnonymisation) {
            builder.header("SP-Anonymous", "*")
        }
        requestHeaders?.let {
            it.forEach { (key, value) ->
                builder.header(key, value)
            }
        }
        return builder.build()
    }

    /**
     * Builds an OkHttp POST request which is ready
     * to be executed.
     * @param request The request where to get the payload to be sent.
     * @param userAgent The user-agent used during the transmission to the collector.
     * @return An OkHttp request object.
     */
    private fun buildPostRequest(request: Request, userAgent: String): okhttp3.Request {
        val reqUrl = uriBuilder.build().toString()
        val reqBody = request.payload.toString().toRequestBody(JSON)
        
        val builder = okhttp3.Request.Builder()
            .url(reqUrl)
            .header("User-Agent", userAgent)
            .post(reqBody)
        if (serverAnonymisation) {
            builder.header("SP-Anonymous", "*")
        }
        requestHeaders?.let {
            it.forEach { (key, value) ->
                builder.header(key, value)
            }
        }
        return builder.build()
    }

    /**
     * Returns a Callable Request Send
     *
     * @param request the request to be sent
     * @return the new Callable object
     */
    private fun getRequestCallable(request: okhttp3.Request): Callable<Int> {
        return Callable { requestSender(request) }
    }

    /**
     * The function responsible for actually sending
     * the request to the collector.
     *
     * @param request The request to be sent
     * @return a RequestResult
     */
    private fun requestSender(request: okhttp3.Request): Int {
        return try {
            Logger.v(TAG, "Sending request: %s", request)
            TrafficStats.setThreadStatsTag(TRAFFIC_STATS_TAG)
            val resp = client?.newCall(request)?.execute()
            resp?.let {
                resp.body?.close()
                return resp.code
            }
            -1
        } catch (e: IOException) {
            Logger.e(TAG, "Request sending failed: %s", e.toString())
            -1
        }
    }

    companion object {
        private const val TRAFFIC_STATS_TAG = 1
        private val DEFAULT_USER_AGENT = String.format(
            "snowplow/%s android/%s",
            BuildConfig.TRACKER_LABEL,
            Build.VERSION.RELEASE
        )
    }
}
